src/security/acl_api.c

```diff
@@ -0,0 +1,450 @@
+	for (i = 0; i < num_aces; i++) {
+		int ace_size = daos_ace_get_size(aces[i]);
+
+		memcpy(current_ace, aces[i], ace_size);
+		current_ace += ace_size;
+	}
+}
+
+/*
+ * Calculates the expected length of the flattened ACE data blob.
+ *
+ * Returns -DER_INVAL if one of the ACEs is NULL.
+ */
+static int
+get_flattened_ace_size(struct daos_ace *aces[], uint16_t num_aces)
+{
+	int	i;
+	int	total_size = 0;
+
+	for (i = 0; i < num_aces; i++) {
+		int len = daos_ace_get_size(aces[i]);
+
+		if (len < 0) {
+			return len;
+		}
+
+		total_size += len;
+	}
+
+	return total_size;
+}
+
+struct daos_acl *
+daos_acl_alloc(struct daos_ace *aces[], uint16_t num_aces)
+{
+	struct daos_acl	*acl;
+	int		ace_len;
+
+	ace_len = get_flattened_ace_size(aces, num_aces);
+	if (ace_len < 0) {
+		/* Bad ACE list */
+		return NULL;
+	}
+
+	sort_aces_by_principal_type(aces, num_aces);
+
+	D_ALLOC(acl, sizeof(struct daos_acl) + ace_len);
+	if (acl == NULL) {
+		/* Couldn't allocate */
+		return NULL;
+	}
+
+	acl->dal_ver = DAOS_ACL_VERSION;
+	acl->dal_len = ace_len;
+
+	flatten_aces(acl->dal_ace, aces, num_aces);
+
+	return acl;
+}
+
+void
+daos_acl_free(struct daos_acl *acl)
+{
+	/* The ACL is one contiguous data blob - nothing special to do */
+	D_FREE(acl);
+}
+
+static bool
+principal_name_matches_ace(struct daos_ace *ace, const char *principal)
+{
+	if (principal == NULL) {
+		/* Nothing to compare */
+		return true;
+	}
+
+	return (strncmp(principal, ace->dae_principal, ace->dae_principal_len)
+			== 0);
+}
+
+static bool
+ace_matches_principal(struct daos_ace *ace,
+		enum daos_acl_principal_type type, const char *principal,
+		size_t principal_len)
+{
+	return	(ace->dae_principal_type == type) &&
+		(ace->dae_principal_len == D_ALIGNUP(principal_len, 8)) &&
+		principal_name_matches_ace(ace, principal);
+}
+
+static bool
+principals_match(struct daos_ace *ace1, struct daos_ace *ace2)
+{
+	return ace_matches_principal(ace1, ace2->dae_principal_type,
+			ace2->dae_principal, ace2->dae_principal_len);
+}
+
+/*
+ * Write the ACE to the memory address pointed to by the pen.
+ * Returns the new pen position.
+ */
+static uint8_t *
+write_ace(struct daos_ace *ace, uint8_t *pen)
+{
+	size_t	len;
+
+	len = daos_ace_get_size(ace);
+	memcpy(pen, (uint8_t *)ace, len);
+
+	return pen + len;
+}
+
+static void
+copy_acl_with_new_ace_inserted(struct daos_acl *acl, struct daos_acl *new_acl,
+		struct daos_ace *new_ace)
+{
+	struct daos_ace	*current;
+	uint8_t		*pen;
+	bool		new_written = false;
+
+	current = daos_acl_get_first_ace(acl);
+	pen = new_acl->dal_ace;
+	while (current != NULL) {
+		if (!new_written && current->dae_principal_type >
+				new_ace->dae_principal_type)  {
+			new_written = true;
+			pen = write_ace(new_ace, pen);
+		} else if (!new_written && principals_match(current,
+					new_ace)) {
+			new_written = true;
+			pen = write_ace(new_ace, pen);
+
+			/* new ACE already in old list - shorten the length */
+			new_acl->dal_len = acl->dal_len;
```

Comment:
Looking below it seems like this means we'd allocate a larger buffer than necessary for new_acl and just truncate the length leaving the full allocation? I don't think it will be a huge issue since the amount of extra space will be small but it is just something I noticed reviewing the add ace realloc call below.

Reply:
Yeah, it's a tradeoff - the alternate option is to walk the list an extra time to see if the principal is already there before getting started. Which we could do, if you like. I guess the question is how long a principal string could be. We don't have a limit set at this point.

Reply:
The more I think about this the more it just seems weird/confusing to me - better to do the real calculation now and optimize later if we find walking the ACLs gets time-consuming.